tests/ui/borrow_interior_mutable_const.rs

```diff
@@ -82,4 +113,6 @@ fn main() {
     assert_eq!(u64::ATOMIC.load(Ordering::SeqCst), 9); //~ ERROR interior mutability
 
     assert_eq!(NO_ANN.to_string(), "70"); // should never lint this.
+
+    let _ = &CELL_REF.0;
```

Comment:
what happens if we manually deref the const here? (wondering if it still lints)

Reply:
That's what this code block is for:

https://github.com/rust-lang/rust-clippy/blob/ce8915c85cebaa30e6846322e0ab44a4f4f9cb65/clippy_lints/src/non_copy_const.rs#L237-L240

In this case (dereferencing the constant before the field access), the code would never reach where I changed since it stops "go further up"; and checks whether `CELL_REF` itself (not `(UnsafeCell<u32>,)`) is a cell (which is not). Actually, that's what happened when I manually dereferenced a raw pointer like this:

```rust
const CELL_PTR: *const (UnsafeCell<u32>,) = std::ptr::null();
let _ = unsafe { &*CELL_PTR }.0;
```

The code above doesn't cause a false positive without this PR. I think the actual description of the false positive the issue reported is that "when implicitly dereferencing a pointer pointing a cell via a field access" because it's the only case in which Clippy assigns `dereferenced_expr` something other than `expr`.